apollo-router/src/query_planner/execution.rs

```diff
@@ -393,6 +400,58 @@ impl PlanNode {
             (value, errors)
         })
     }
+
+    /// This is a reimplementation of `ValueExt::deep_merge` which takes type hierarchies into account
+    /// when merging response objects. If we are merging two instances of `__typename`, we check the
+    /// schema to make sure we do not overwrite a subtype with an interface or parent type. Otherwise,
+    /// the data associated with the interface/parent type is removed from the response later on.
+    fn type_aware_deep_merge(schema: &Schema, target: &mut Value, source: Value) {
```

Comment:
Should this be added directly in `ValueExt` and its `Value` implementation instead? 

Also, there is another usage of `deep_merge` happening in [`execute_selection_set.rs`](https://github.com/apollographql/router/blob/6c32fd2b8fb19ecc4b3ade2bfd09cd9ac8508992/apollo-router/src/query_planner/selection.rs#L178). Does that need to be typed aware as well?

Reply:
My understanding is that we don't need a type-aware merge in `execute_selection_set()`. Type-aware merging needs to happen whenever we're merging fetched data with other fetched data. In the case of `execute_selection_set()`, it's being called to execute selection sets against already-merged fetched data, for the purposes of computing variable values. That selection set execution may need to merge to handle inline fragments, but there won't ever be disagreements on the value of `__typename` in that merging process.

Reply:
> Should this be added directly in `ValueExt` and its `Value` implementation instead?

Yeah, I think that would be a better place for it.

Reply:
Moved this to `ValueExt` and updated the `execute_selection_set` implementation as well.

> Do you know whether or not there are any performance implication for this change?

The latest perf test results look normal to me, but let me know if there's a specific part you're worried about.